In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('devices_v3.csv')
df.head()

id    brand                   name  \
0   1  INFINIX            Infinix S3X   
1   2  INFINIX         Infinix Hot 6X   
2   3  INFINIX  Infinix Note 5 Stylus   
3   4  INFINIX     Infinix Smart 2 HD   
4   5  INFINIX         Infinix Note 5   

                                                 url   type  \
0      https://www.gsmarena.com/infinix_s3x-9420.php  phone   
1   https://www.gsmarena.com/infinix_hot_6x-9423.php  phone   
2  https://www.gsmarena.com/infinix_note_5_stylus...  phone   
3  https://www.gsmarena.com/infinix_smart_2_hd-94...  phone   
4   https://www.gsmarena.com/infinix_note_5-9417.php  phone   

  network_technology     network_types  sensors_count  \
0                 4G  GSM / HSPA / LTE            5.0   
1                 4G  GSM / HSPA / LTE            4.0   
2                 4G  GSM / HSPA / LTE            5.0   
3                 3G        GSM / HSPA            4.0   
4                 4G  GSM / HSPA / LTE            5.0   

                                       sensors_types  length  ...  sim_count  \
0  fingerprint / accelerometer/ gyro/ proximity/ ...   156.7  ...        2.0   
1    fingerprint / accelerometer/ proximity/ compass   157.0  ...        2.0   
2  fingerprint / accelerometer/ gyro/ proximity/ ...   158.0  ...        2.0   
3    accelerometer/ proximity/ compass/ fingerprint    156.9  ...        2.0   
4  fingerprint / accelerometer/ gyro/ proximity/ ...   158.0  ...        2.0   

   os_type  os_version  price  released_date  cpu_core  ram  internal_storage  \
0  Android         8.1    NaN         2018.0       8.0  3GB              32GB   
1  Android         8.1    NaN         2018.0       4.0  2GB              16GB   
2  Android         8.1    NaN         2018.0       8.0  4GB              32GB   
3  Android         8.1    NaN         2018.0       4.0  1GB              16GB   
4  Android         8.1   90.0         2018.0       8.0  3GB              32GB   

  fan_count  view_count  
0        10         485  
1        11         736  
2        31         524  
3         1         335  
4        35         871  

[5 rows x 31 columns]

In [3]:
df['sim_count'].value_counts()

sim_count
1.0    2709
2.0    2400
0.0     590
3.0       7
4.0       2
Name: count, dtype: int64

In [4]:
df_clean = df.dropna(subset=['price'])
df_clean.head()

id    brand                 name  \
4    5  INFINIX       Infinix Note 5   
25  26  INFINIX     Infinix Note 11i   
26  27  INFINIX  Infinix Smart 5 Pro   
27  28  INFINIX     Infinix Note 11s   
28  29  INFINIX      Infinix Note 11   

                                                  url   type  \
4    https://www.gsmarena.com/infinix_note_5-9417.php  phone   
25  https://www.gsmarena.com/infinix_note_11i-1121...  phone   
26  https://www.gsmarena.com/infinix_smart_5_pro-1...  phone   
27  https://www.gsmarena.com/infinix_note_11s-1120...  phone   
28  https://www.gsmarena.com/infinix_note_11-11201...  phone   

   network_technology     network_types  sensors_count  \
4                  4G  GSM / HSPA / LTE            5.0   
25                 4G  GSM / HSPA / LTE            5.0   
26                 4G  GSM / HSPA / LTE            3.0   
27                 4G  GSM / HSPA / LTE            5.0   
28                 4G  GSM / HSPA / LTE            5.0   

                                        sensors_types  length  ...  sim_count  \
4   fingerprint / accelerometer/ gyro/ proximity/ ...   158.0  ...        2.0   
25  fingerprint / accelerometer/ gyro/ proximity/ ...   173.2  ...        2.0   
26             fingerprint / accelerometer/ proximity   164.4  ...        2.0   
27  fingerprint / accelerometer/ gyro/ proximity/ ...   173.1  ...        2.0   
28  fingerprint / accelerometer/ gyro/ proximity/ ...   164.4  ...        2.0   

    os_type  os_version  price  released_date  cpu_core  ram  \
4   Android         8.1   90.0         2018.0       8.0  3GB   
25  Android          11  140.0         2021.0       8.0  4GB   
26  Android          11   70.0         2021.0       8.0  2GB   
27  Android          11  180.0         2021.0       8.0  4GB   
28  Android          11  100.0         2021.0       8.0  4GB   

    internal_storage fan_count  view_count  
4               32GB        35         871  
25              64GB        19         267  
26              32GB        10         311  
27              64GB        44         607  
28              64GB        61           1  

[5 rows x 31 columns]

### First Question

#### 1.Do mobile devices with different types of SIM cards have different prices?

##### Null Hypothesis (H0): There is no difference in the average price of mobile devices across different SIM card configurations.

##### Alternative Hypothesis (H1): There is a difference in the average price of mobile devices across different SIM card configurations.

In [5]:
df_clean['sim_count'] = df_clean['sim_count'].astype('category')
df_clean['sim_count']

/var/folders/3g/9l1ty6zs7f95d8vdxzyb6m5h0000gn/T/ipykernel_22171/539397305.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean['sim_count'] = df_clean['sim_count'].astype('category')


4       2.0
25      2.0
26      2.0
27      2.0
28      2.0
       ... 
5667    2.0
5668    1.0
5671    1.0
5685    1.0
5687    1.0
Name: sim_count, Length: 4364, dtype: category
Categories (5, float64): [0.0, 1.0, 2.0, 3.0, 4.0]

In [6]:
from scipy.stats import f_oneway

In [7]:
groups = df_clean.groupby('sim_count')['price'].apply(list)
groups

/var/folders/3g/9l1ty6zs7f95d8vdxzyb6m5h0000gn/T/ipykernel_22171/2781704084.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  groups = df_clean.groupby('sim_count')['price'].apply(list)


sim_count
0.0    [80.0, 100.0, 80.0, 130.0, 110.0, 100.0, 80.0,...
1.0    [250.0, 130.0, 80.0, 120.0, 130.0, 60.0, 90.0,...
2.0    [90.0, 140.0, 70.0, 180.0, 100.0, 120.0, 100.0...
3.0                     [90.0, 90.0, 100.0, 60.0, 120.0]
4.0                                               [90.0]
Name: price, dtype: object

In [8]:
anova_result = f_oneway(*groups)
print(f"F-statistic: {anova_result.statistic}, p-value: {anova_result.pvalue}")

F-statistic: 41.171287626252685, p-value: 6.355681231524028e-34


##### Assume α=0.05, as p-value indicates it is less than that which leads to the rejection of null hypothesis.

#### 2.Do mobile devices with different sizes have different prices?

##### Let's consider "weight" as a proxy for size.
##### Assuming that heavier devices might be larger or more feature-rich, potentially affecting their price.

In [9]:
df_clean = df.dropna(subset=['weight', 'price'])
df_clean.head()

id    brand             name  \
4    5  INFINIX   Infinix Note 5   
28  29  INFINIX  Infinix Note 11   
31  32  INFINIX  Infinix Hot 11s   
33  34  INFINIX   Infinix Hot 11   
34  35  INFINIX   Infinix Zero X   

                                                  url   type  \
4    https://www.gsmarena.com/infinix_note_5-9417.php  phone   
28  https://www.gsmarena.com/infinix_note_11-11201...  phone   
31  https://www.gsmarena.com/infinix_hot_11s-11162...  phone   
33  https://www.gsmarena.com/infinix_hot_11-11161.php  phone   
34  https://www.gsmarena.com/infinix_zero_x-11097.php  phone   

   network_technology     network_types  sensors_count  \
4                  4G  GSM / HSPA / LTE            5.0   
28                 4G  GSM / HSPA / LTE            5.0   
31                 4G  GSM / HSPA / LTE            5.0   
33                 4G  GSM / HSPA / LTE            5.0   
34                 4G  GSM / HSPA / LTE            5.0   

                                        sensors_types  length  ...  sim_count  \
4   fingerprint / accelerometer/ gyro/ proximity/ ...   158.0  ...        2.0   
28  fingerprint / accelerometer/ gyro/ proximity/ ...   164.4  ...        2.0   
31  fingerprint / accelerometer/ gyro/ proximity/ ...   168.9  ...        2.0   
33  fingerprint / accelerometer/ gyro/ proximity/ ...   164.7  ...        2.0   
34  fingerprint / accelerometer/ gyro/ proximity/ ...   164.1  ...        2.0   

    os_type  os_version       price  released_date  cpu_core  ram  \
4   Android         8.1   90.000000         2018.0       8.0  3GB   
28  Android          11  100.000000         2021.0       8.0  4GB   
31  Android          11  119.967978         2021.0       8.0  4GB   
33  Android          11  110.000000         2021.0       8.0  4GB   
34  Android          11  320.000000         2021.0       8.0  8GB   

    internal_storage fan_count  view_count  
4               32GB        35         871  
28              64GB        61           1  
31              64GB        64           1  
33              64GB        26           1  
34             128GB        61         688  

[5 rows x 31 columns]

In [11]:
from scipy.stats import linregress

In [12]:
weight = df_clean['weight']
price = df_clean['price']

slope, intercept, r_value, p_value, std_err = linregress(weight, price)
print(f"Slope (Coefficient): {slope}")
print(f"Intercept: {intercept}")
print(f"R-squared: {r_value**2}")
print(f"P-value: {p_value}")

Slope (Coefficient): 0.5643015701019873
Intercept: 138.90973262726607
R-squared: 0.03834108602165969
P-value: 1.3016983188211696e-37


##### Slope Value  
for each additional unit of weight, 
The price of the mobile device increases by about 27.82 units.This indicates a positive relationship between weight and price.
##### Intercept
No meaning! cause weight of 0 for a phone is not practically feasible in this context.
##### R-squared
Is closed to 0 so this means that weight, on its own, is not a good predictor of price for mobile devices in this dataset. There are likely other factors that have a much stronger influence on the price.
##### P-value
It's below the alpha=0.05 thus this suggests that there is a statistically significant relationship between weight and price.

### Second Question

#### Null Hypothesis (H0): There is no difference in the mean PPI density between large and small mobile devices.
#### Alternative Hypothesis (H1): There is a difference in the mean PPI density between large and small mobile devices.

In [13]:
df['size_category'] = df['screen_size'].apply(lambda x: 'large' if x >= 7.00 else 'small')

In [14]:
from scipy.stats import ttest_ind

In [15]:
large_devices = df[df['size_category'] == 'large']['ppi'].dropna()
small_devices = df[df['size_category'] == 'small']['ppi'].dropna()


In [16]:
t_stat, p_value = ttest_ind(large_devices, small_devices)
print(f"T-statistic: {t_stat}, P-value: {p_value}")


T-statistic: -7.852605710294054, P-value: 4.893430093041532e-15


##### T-statistics
Indicates significant difference between the mean PPI densities of large and small mobile devices
##### P-value
It's below the alpha=0.05 and extremely small thus this suggests that null hypothesis is not true.
The results strongly suggest that there is a statistically significant difference in PPI density between the two groups, with small devices likely having higher PPI densities compared to larger ones.

### Third Question

#### Null Hypothesis (H0): There is no difference in the mean weight of Android and iOS mobile devices.
#### Alternative Hypothesis (H1): There is a difference in the mean weight of Android and iOS mobile devices.

In [17]:
android_weights = df[df['os_type'] == 'Android']['weight'].dropna()
ios_weights = df[df['os_type'] == 'iOS']['weight'].dropna()

In [18]:
t_stat, p_value = ttest_ind(android_weights, ios_weights)
print(f"T-statistic: {t_stat}, P-value: {p_value}")

T-statistic: -8.27257493916851, P-value: 1.8867397072458762e-16


##### T-statistics
The magnitude of the T-statistic reflects a substantial difference between the two groups.
##### P-value
It's below the alpha=0.05 and extremely small thus this suggests that null hypothesis is not true.

### Fourth Question

#### Null Hypothesis (H0): The mean battery capacity per hour is the same across the different combinations of screen size classes and brands (Apple, Samsung, Xiaomi).
#### Alternative Hypothesis (H1): At least one group's mean battery capacity per hour differs from the others.

In [19]:
screen_size_threshold = 7.00
df['screen_size_class'] = df['screen_size'].apply(lambda x: 'large' if x >= screen_size_threshold else 'small')
df['screen_size_class']

0       small
1       small
2       small
3       small
4       small
        ...  
5704    small
5705    small
5706    small
5707    small
5708    small
Name: screen_size_class, Length: 5709, dtype: object

In [20]:
df_filtered = df[df['brand'].isin(['APPLE', 'SAMSUNG', 'XIAOMI'])]
df_filtered.head()

id  brand                              name  \
533  534  APPLE             Apple iPad mini Wi-Fi   
534  535  APPLE  Apple iPad mini Wi-Fi + Cellular   
536  537  APPLE                Apple iPad 4 Wi-Fi   
537  538  APPLE                    Apple iPhone 5   
538  539  APPLE     Apple iPad 3 Wi-Fi + Cellular   

                                                   url    type  \
533  https://www.gsmarena.com/apple_ipad_mini_wi_fi...  tablet   
534  https://www.gsmarena.com/apple_ipad_mini_wi_fi...  tablet   
536  https://www.gsmarena.com/apple_ipad_4_wi_fi-50...  tablet   
537   https://www.gsmarena.com/apple_iphone_5-4910.php   phone   
538  https://www.gsmarena.com/apple_ipad_3_wi_fi_+_...  tablet   

    network_technology                   network_types  sensors_count  \
533                 2G        No cellular connectivity            3.0   
534                 4G         GSM / CDMA / HSPA / LTE            3.0   
536                 2G        No cellular connectivity            3.0   
537                 4G  GSM / CDMA / HSPA / EVDO / LTE            4.0   
538                 4G  GSM / CDMA / HSPA / EVDO / LTE            3.0   

                               sensors_types  length  ...  os_version  price  \
533             accelerometer/ gyro/ compass   200.0  ...           6  200.0   
534             accelerometer/ gyro/ compass   200.0  ...           6  300.0   
536             accelerometer/ gyro/ compass   241.2  ...           6  350.0   
537  accelerometer/ gyro/ proximity/ compass   123.8  ...           6  340.0   
538             accelerometer/ gyro/ compass   241.2  ...         5.1  400.0   

     released_date  cpu_core  ram  internal_storage  fan_count  view_count  \
533         2012.0       2.0  NaN               NaN        221           5   
534         2012.0       2.0  NaN               NaN        150           5   
536         2012.0       2.0  1GB              16GB        162           2   
537         2012.0       2.0  1GB              16GB       1973          56   
538         2012.0       2.0  1GB              16GB         93           4   

    size_category  screen_size_class  
533         large              large  
534         large              large  
536         large              large  
537         small              small  
538         large              large  

[5 rows x 33 columns]

In [21]:
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [22]:
model = ols('battery_capacity ~ C(screen_size_class) * C(brand)', data=df_filtered).fit()

In [23]:
anova_results = sm.stats.anova_lm(model, typ=2)
print(anova_results)

                                     sum_sq      df            F  \
C(screen_size_class)           2.985135e+09     1.0  1380.791455   
C(brand)                       1.543471e+09     2.0   356.970718   
C(screen_size_class):C(brand)  1.106721e+08     2.0    25.596002   
Residual                       3.779004e+09  1748.0          NaN   

                                      PR(>F)  
C(screen_size_class)           3.019696e-223  
C(brand)                       1.010376e-130  
C(screen_size_class):C(brand)   1.105256e-11  
Residual                                 NaN  


### Fifth Question

#### Null Hypothesis (H0): There is no difference in the mean price of mobile devices between the different screen size classes (large vs. small).
#### Alternative Hypothesis (H1): There is a difference in the mean price of mobile devices between the different screen size classes (large vs. small).

In [42]:
model = ols('price ~ C(screen_size_class)', data=df).fit()
anova_results = sm.stats.anova_lm(model, typ=2)

print(anova_results)

                            sum_sq      df         F    PR(>F)
C(screen_size_class)  4.029644e+10     1.0  3.665762  0.055607
Residual              4.794994e+13  4362.0       NaN       NaN


The p-value of 0.055607 is just above the common significance level threshold of 0.05 so evidence is not strong enough to reject the null hypothesis.

### Sixth Question

#### Null Hypothesis (H0): The mean weight of large mobile devices is less than or equal to the mean weight of small mobile devices. 
#### Alternative Hypothesis (H1): The mean weight of large mobile devices is greater than the mean weight of small mobile devices. 

In [24]:
large_weights = df[df['size_category'] == 'large']['weight'].dropna()
small_weights = df[df['size_category'] == 'small']['weight'].dropna()

In [25]:
t_stat, p_value_two_sided = ttest_ind(large_weights, small_weights, equal_var=False)

p_value_one_sided = p_value_two_sided / 2

if t_stat > 0 and p_value_one_sided < 0.05:
    print(f"Evidence supports the alternative hypothesis with t-statistic = {t_stat}, one-sided p-value = {p_value_one_sided}")
else:
    print(f"No evidence to support the alternative hypothesis with t-statistic = {t_stat}, one-sided p-value = {p_value_one_sided}")

Evidence supports the alternative hypothesis with t-statistic = 40.88793412131005, one-sided p-value = 4.009929556855361e-159


With a p-value far below any conventional significance level (e.g., 0.05, 0.01), we can confidently reject the null hypothesis